<figure>
  <IMG SRC="https://raw.githubusercontent.com/mbakker7/exploratory_computing_with_python/master/tudelft_logo.png" WIDTH=250 ALIGN="right">
</figure>

# 1. Variables, operators and functions. 


## 1.1 First Python Script

So, it is time for your first Python script. It is located beneath. Go ahead, run it.


In [1]:
# My First Python Script
message = 'Hello world!'
print(message)

Hello world!



`````{admonition} Let's break it down

First line: <b><code># My First Python Script</code></b> is a <b>comment</b>, which is used just to explain and/or put useful information about the code next to it. If you need to create a comment — just type a <b><code>#</code></b> symbol and write your text after it. The interpreter does nothing with it — the comments are there just as useful information for you or another reader.

Second line: <b><code>message = 'Hello world!'</code></b> creates a <b>variable</b> called <i> message </i> and assigns the text literal (string) <i>Hello world!</i> to it, by using the operator <i>=</i> (equal sign). Variables are used to store all data you use in your code.

Third line: <b><code>print(message)</code></b> calls the <b>function</b><code>print()</code> and passes the variable <i>message</i> to it. A function is just a set of encapsulated code, which is tailored to perform a certain action. This specific function outputs the content of everything you pass to it. Since the variable <i>message</i> had a small line of text in it, the function <b><code>print()</code></b> outputs that message.

This script is quite primitive and simple but it represents the main idea of programming in Python. You have data (which is stored in variables) and you perform an operation on it: by using the inbuilt <b><code>print()</code></b> function. In addition, you could create your own functions. 

`````

## 1.2 Python Variables 

One of the most powerful features of a programming language is the ability to create and manipulate
variables. A variable is a labeled storage that refers to a value. They are usually used in a way to make the code more readable, allowing reusability of your code.<br><br>As it was mentioned previously, Python is a high-level programming language. For that, it uses concepts of <b>Class</b> and <b>Object</b>. In short, a <b>Class</b> is a defined model or a data structure, it describes how the data of that class can be described and how it can be manipulated, and an <b>Object</b> or <b>Instance</b> is its realization. In other words, think about your favorite meal: the recipe is the <b>Class</b> of your meal and when you decide to use the recipe and cook it — you create an <b>Object</b> of that <b>Class</b>. Variables are the way how objects are stored and processed.


### Rules for variable names
* names can not start with a number
* names can not contain spaces, use _ instead
* names can not contain any of these symbols:

      :'",<>/?|\!@#%^&*~-+
      
::: {note}     
* it's considered best practice ([PEP8](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names)) that names are lowercase with underscores
* avoid using Python built-in keywords like `list` and `str`
::: 

### Assigning Variables

Variable assignment follows <b><code>variable_name = value</code></b>, where a single equal sign $=$ is an <b>assignment operator</b>. More on operators will be covered in the next section. Let's see a few examples of how we can do this.

In [2]:
# Let's create an object called "a" and assign it the number 5
a = 5

Now if I call `a` in my Python script, Python will treat it as the number $5$.

In [3]:
# Adding the objects
a + a

10

What happens on reassignment? Will Python let us write over it?

In [4]:
# Reassignment
a = 20

In [5]:
# Check
print(a)

20


Yes! Python allows you to overwrite assigned variable names. We can also use the variables themselves when doing the reassignment.

Since <b><code>a = 20</code></b> was the last assignment to our variable <b><code>a</code></b>, you can keep using <b><code>a</code></b> in place of the number <b><code>20</code></b>:

In [6]:
a + a

40

Instead of writing <b><code>a+a</code></b>, Python has a built-in shortcut for these simple operations.<br><br>You can add, subtract, multiply and divide numbers with reassignment using <b><code>+=</code></b>, <b><code>-=</code></b>, <b><code>*=</code></b>, and <b><code>/=</code></b>, respectively.

In [7]:
a += 10

The above code will add **`10`** to the variable **`a`** every time you run that cell. 

Try it yourself, run it a few times and then run the below cell to see what's the value of **`a`**
.

In [8]:
print(a)

30


Below an example of a code that will double **`a`** every time that you run that cell.

In [9]:
a *= 2
print(a)

60


### Determining variable type with <b><code>type()</code></b>

You can check what type of object is assigned to a variable using Python's built-in <b><code>type()</code></b> function. Common data types include:


* **int** (for integer numbers)
* **float** (for floating point / all real numbers)
* **str** (for string/text)
* **bool** (for Boolean True/False)
* **list**
* **tuple**
* **dict** 
* **set**


::: {warning}
Always check the type of your variables when debugging!
:::

Below a few examples:

In [10]:
type(a)

int

In [11]:
float_var = 3.1415
type(float_var)

float

In [12]:
a = 0.3
b = 0.2
c = a-b
print(c)

0.09999999999999998


You probably noticed that Python wrote $0.09999999999999998$ instead of $1$ when calculating $0.3 - 0.2$. Ignore it for now, this is explained later in this Notebook.

In [13]:
type(1<2)

bool

Boolean variables can only take on two values: <b><code>True</code></b> or <b><code>False</code></b>. They are often used to check conditions.

In [14]:
1<2

True

In [15]:
# the variable from the first script
type(message)

str

<b>Strings</b> are variables represented in between <b><code>' '</code></b> or <b><code>" "</code></b>.<br><br>They are a <b>sequence</b> of values, therefore you are able to access and manipulate every character individually.<br><br>This is done with the bracket operator <b><code>[]</code></b>, which works as an <b>index</b>.<br><br>Let's take a look at our first variable from this notebook: <b><code>message</code></b>.

In [16]:
message

'Hello world!'

In [17]:
message[1]

'e'

<b>Nani???</b>

Why index <b><code>[1]</code></b> gave us the second letter? For most people, the first letter of the sentence <b><code>Hello world!</code></b> is <b><code>H</code></b> not <b><code>e</code></b>.

So.. what happened?

In Python, indexing starts at **`[0]`**. **`H`** is the *zero-th* character of **`Hello world!`**.

In [18]:
message[0]

'H'

You can also access the last value of a string using the index <b><code>[-1]</code></b>, the before-last using <b><code>[-2]</code></b> and so forth.. This will turn out to be very useful!

In [19]:
message[-1]

'!'

Strings are also immutable. You cannot reassign a new value for one of the characters. You will have to create a new string for that.

In [20]:
message[0] = 'J'

TypeError: 'str' object does not support item assignment

You can also add (i.e. concatenate) strings and characters. But it will create a new string, it will not modify the old one (since they are immutable).

In [21]:
message + message

'Hello world!Hello world!'

In [22]:
message[0]+message[1]+message[2]+message[3]

'Hell'

A segment of a string is called a <b>slice</b>. Selecting a slice is similar to selecting a character. Using the operator <b><code>:</code></b> you select the first value that you want to <b>get</b>, and the first value you want to <b>leave out</b> of the slice, for example:<br><br>

Let's say we want to write the word <b>Hell</b> using our variable <b><code>message</code></b>, without having to type as much as above.

1. Which letter is the first we want to <b>get</b>? **`H`**, which has index **`[0]`**.

2. Which letter is the first we want to <b>leave out</b>? **`o`**, which has index **`[4]`**. So...

In [23]:
message[0:4]

'Hell'

One can also use our variable <code>message</code> to write the sentence <b><code>Hold door</code></b>.

In [25]:
message[0]+message[4]+message[-3:-1]+message[5]+message[-2]+2*message[4]+message[-4]

'Hold door'

::: {note}
Real life example: Analyzing satellite data<br><br></b>Sentinel data title is formatted as: <b>S1A_IW_SLC__1SDV_20181205T015821_20181205T015851_024884_02BD8C_8700</b><br>
    where each part means something, S1A means Sentinel-1A, IW means Interferometric Wide Swath<br> 20181205T015821 is a date, 2018 12 05, at 01h58m21s, etc.<br><br> Therefore, being able to manipulate this string is fundamental in order to organize and select satellite data. 
::: 

### Dynamic Typing

The way the computer is able to use the stored data becomes an attribute of the stored data. This attribute is called a <b>data type</b>. Python uses <b>dynamic typing</b>, meaning you can also reassign variables to different data types. This makes Python very flexible in assigning data types; it differs from other languages that are <b>statically typed</b>.

#### Pros and Cons of Dynamic Typing
##### Pros of Dynamic Typing
* very easy to work with
* faster development time

##### Cons of Dynamic Typing
* may result in unexpected bugs!
* you need to be aware of **`type()`**.

In [26]:
a = 5
print('Type of a is =',type(a))
a = 'string'
print('Type of a is =',type(a))

Type of a is = <class 'int'>
Type of a is = <class 'str'>


See, now **`a`** is no longer an **`int`** type but a **`str`** type

### Casting types

Sometimes you want to change the type of a variable. For example, there is no point in arithmetically adding a number to a string. Or, when estimating the amount of oil wells, those should be integers (how would you build $2.5$ wells?). These two problems can be sometimes solved with casting.<br><br><b>Casting</b> is a procedure of changing variables type. Actually, you create a new variable with the requested data type using the variable you want to alter.

Examples are shown below.
    

In [27]:
string_number = '123'
print(string_number, type(string_number))

integer_number = int(string_number)
print(integer_number, type(integer_number))

123 <class 'str'>
123 <class 'int'>


As you can see, both variables look the same in the output but their type now is different. Because of that, the cell below will result in an error.

In [28]:
string_number + 5

TypeError: can only concatenate str (not "int") to str

But the next cell will run normally.

In [29]:
integer_number + 5

128

### How do computers read variables?

Computers and humans have different views on data. We, fellow humans, can easily read poetry and enjoy music. Computers are limited in that sense and they can operate on a really primitive level. They cannot even read normal text, for a computer, everything is represented with numbers. One of the popular text encodings <b>ASCII</b> has a table, where every symbol there is an integer number assigned to it. Letter 'a', for instance, has a value of $97$, and letter 'A' has a value of $65$. The value of 'B' is $66$ and if we want to take the letter before 'B' we just calculate $66 - 1$. You don't have to open an alphabet and, therefore, it is easier to work with numbers than with letters.<br><br>But even for numbers, computers have their own approach. On a really low level, all numbers are stored and operated in their binary form. Number $5$ becomes $101$, because $5 = 1\cdot2^2 + 0 \cdot2^1 + 1\cdot2^0$. And since all symbols are also represented with numbers, it is not hard for computers to operate with them. But what about the floating numbers? How do you represent $1.5$ or $\pi$? Well, computers try to represent them with binary code, which introduces some problems. In order to represent a fraction, they need an infinite amount of bits, to obtain the precision we use in math. But that's impossible! In other words, small errors will always be present in calculations as simple as $0.3 - 0.2$. From your math classes you know that the result of $0.3 - 0.2 = 0.1$, not $0.10001$ or $0.09999$. But try to run the cell below and see the result.

In [30]:
0.3 - 0.1

0.19999999999999998

It is not what you would expect to see, right? The result has an error of $\approx 10^{-15}$. In most cases, it can be neglected but be careful when comparing <b><code>float</code></b> and <b><code>int</code></b> numbers, as shown below.

In [31]:
0.2 == 0.3-0.1

False

Indeed, $0.2 \neq 0.19999999999999998$.

::: {note}
Within this course, in the end of each section you'll find exercises related to the subject that was just covered. There are three types of exercises: normal, fixing and searching.

Normal exercises are straight forward exercises that you should be able to solve without much trouble.
Fixing exercises are exercises where some piece of code is already written and you need to debug it (find the root of the error and fix it).

Searching exercises are exercises that purposefully incorporate subjects that were not covered yet, in an attempt to encourage you to try and solve issues you haven't learned about yet.
:::

## 1.3 Python Operators

Python <b>operators</b> are used to perform operations on variables and values. They are symbols that represent a form of computation; think of addition or multiplication. The value to which this computation is applied to is called the <i>'operand'</i>. Most of the common operators you will recognize from mathematics.

<h3> Arithmetic Operators </h3>

| Math sign | Python sign | name |
| :-: | :-: |:-:|
| + | + | addition |
| - | - | subtraction |
| * | * | multiplication |
| / | / | division |
| ^ | ** | exponentiation |
| mod | % | modulus |
|  | // | floor division |


<br>Most of the mathematical symbols stay the same when transforming a piece of mathematics to Python. Note that the exponentiation sign is a double multiplication sign!<br><br>
The last two operators, modulus and floor division, can be defined as the following:<br>
- modulus: return the remainder of a division<br>
- floor division: returns the integer/whole part of the division


Now we will provide some small examples

a. multiply 4 by 3 and then add 2

b. 2 to the power of 4 plus 1

c. take the modulus of 352 over 23

d. the floor division of 352 over 23




In [32]:
a = 2 + 4 * 3
print(a)

14


In [33]:
b = 2**4 + 1
print(b)

17


In [34]:
c = 352 % 23
print(c)

7


Explanation: $352 = 15 \times 23 + 7$, therefore the **modulus operator** returns the value $7$.

In [35]:
d = 352 // 23
print(d)

15


Explanation: $352 = 15 \times 23 + 7$, therefore the **floor division operator** returns the value $15$.

Besides making sure that you use the right operators when writing mathematical functions, it is also important that you pay attention to the order of operators. When not done right, this can cause huge changes in the outcome. Therefore, when writing out large equations it is easier to use parentheses or split it into multiple variables. e.g.:

$$y = x\tan\theta - \frac{1}{2v_0^2}\frac{g x^2}{\cos^2\theta} + y_0$$

You could split this equation into four distinct variables:

1. var_1 $ = x\tan\theta$
2. var_2 $= \frac{1}{2v_0^2}$
3. var_3 $= \frac{g x^2}{\cos^2\theta}$
4. var_4 $= y_0$

And then re-write it as ``y = var_1 - (var_2 * var_3) + var_4``

In [36]:
parenthesis_before = (2 + 4) * 3
print('With parenthesis before =',parenthesis_before)
parenthesis_after = 2 + (4 * 3)
print('With parenthesis after =',parenthesis_after)

With parenthesis before = 18
With parenthesis after = 14


### Comparison Operators

In Python, you often want to compare a value with another. For that, you use comparison operators.

| Math sign | Python sign | Meaning |
| :-: | :-: | :-: |
| $=$ | ``==`` | Equal to |
| $>$ | ``>`` | Greater than |
| $>$ | ``<`` | Less than |
| $\geqslant$ | ``>=`` | Greater than or equal to |
| $\leqslant$ | ``<=`` | Less than or equal to |
| $\neq$ | ``!=`` | Not equal to |

#### Checking if a value corresponds to the set conditions

Check if the the variable **`num`** satisfies the set condition.

In [37]:
num = 6
print(num > 2)

True


If the value does not satisfy the condition the system will return <b><code>False</code></b>

In [38]:
print(num > 7)

False



### Logical & Identity Operators
 
 |sign|description|
 |:-:|:-:|
 |and|returns True if both statements are true|
 |or|return True if at least 1 statements is true|
 |not|reverse of the results; returns False if the statement is True|
 |is|returns True if both variables are the same object|
 |is not|returns True if both variables are not the same object|
 |in|returns True if a sequence with the specified value is present in the object|
 |in not|returns True if a sequence with the specified value is not present in the object|

#### <code>and</code></b> statement

By using the <b><code>and</code></b> statement you can set multiple conditions for the system to return. This can be seen as setting a boundary condition for a mathematical function.

In [39]:
num = 5
print(num > 4 and num < 8)

True


#### checking if a value appears in an object

Suppose we have a string "sandstone", we can check if a value is present within the string through the following lines of code.

In [ ]:
rock_type = "sandstone"
print("sand" in rock_type)

### <code>is</code></b> and <b><code>==</code></b> operators

The <b><code>is</code></b> operator deserves a little more explanation since it can be easily confused with the <b><code>==</code></b> operator. The <b><code>is</code></b> statement does not compare the value of a variable but simply checks if two variables are the same object. On the other hand, <b><code>==</code></b> checks if the values of different variables are the same. In the underneath piece of code this is shown quite clearly. Although the values of the variables are the same, their <b><code>type</code></b> is not. Therefore, when compared using the <b><code>is</code></b> operator, it returns <b><code>False</code></b>.

In [40]:
x = 2.0
y = 2

print(type(x),type(y),x is y)
print(x == y)

<class 'float'> <class 'int'> False
True


## 1.4 Python Functions

A <b>function</b> is a collection of code that is assigned to a specific name. You have already seen some built-in Python functions, such as <b><code>print()</code></b>. Using functions is useful because it allows you to run the same code again without having to type it a second time.

Below are some examples of common built-in Python functions and what they do:


* ``print()``: Prints input to screen
* ``type()``: Returns the type of the input
* ``abs()``: Returns the absolute value of the input
* ``min()``: Returns the minimum value of the input. (input could be a list, tuple, etc.)
* ``max()``: Same as above, but returns the maximum value
* ``sum()``: Returns the sum of the input (input could be a list, tuple, etc.)


But the story doesn't end at built-in functions. You can also write your own functions!

### How to write a function

To write a function, you must first define it. This is done by using the <b><code>def</code></b> statement. Then, you name the function and add, in the parentheses, which variables this function takes as an input, followed by a colon. The colon tells Python you are going to define the <b>function body</b> next (the part of the function that actually does the computation). As shown below:<br><br>

<center><code>def function_name(input1, input2,...):<br>function_body<br>...<br>...</code></center>



The <b><code>calculate_circle_area(r)</code></b> function below defines <b><code>pi</code></b> as a variable and then computes the area of a circle, which is stored in the variable <b><code>area</code></b>. Finally, it uses the <b><code>return</code></b> statement to output the <b><code>area</code></b> back to you. Once you have defined the function, you can then <b>call</b> it by typing the function name, and  the inputs in the parenthesis. For example, calling: <b><code>print("Hello World!")</code></b>, prints the string <b><code>"Hello World!"</code></b>.

<b>Indentation</b> <br>

It is worth noting that the function body should be <b>indented</b>. This is how Python sees what piece of code is inside another code. An indented line of code is a line that starts with whitespace. You can do this by using the tab key on your keyboard.

::: {note}
Inputs of functions are more often called <b>arguments</b>.
:::

### indentation of code within functions

Let's say you'd want to compute the area of a circle, but you don't want to calculate $\pi r^2$ the entire time. Then you can write a couple lines of code to do it for you, and wrap it up into a function, like the one below:


In [41]:
def calculate_circle_area(r):
    pi = 3.141592653589793
    area = pi*(r**2)
    return area

This function is called <b><code>calculate_circle_area(r)</code></b>, and takes the value <b><code>r</code></b> as an argument.

Functions can have multiple arguments, for example:

In [42]:
def calculate_rectangle_area(a, b):
    area = a * b
    return area

print('Area of my rectangle is', calculate_rectangle_area(4, 6), 'cm^2')

Area of my rectangle is 24 cm^2


In the cell above, the **`calculate_rectangle_area(a, b)`** function takes $2$ arguments, $a$ and $b$. 

The built-in function **`print()`** takes $3$ arguments:<br>
the <b>string</b> <b><code>'Area of my rectangle is'</code></b>, the output of <b><code>calculate_rectangle_area(a, b)</code></b>, and another string <b><code>'cm^2'</code></b>.

There are better ways to use the built-in <b><code>print()</code></b> function when writing long sentences that have variables in between. For example:

In [46]:
print('Area of my rectangle is {} cm^2 and the area of my circle is {} cm^2'. \
      format(calculate_rectangle_area(4,6), calculate_circle_area(4)))

Area of my rectangle is 24 cm^2 and the area of my circle is 50.26548245743669 cm^2


If a line in your code extends the width of your page, you can use a **\\** at the point where you want to break the line, as shown above.

Note that the variables (separated by commas) called inside the <b><code>.format()</code></b> will appear, in order, where the <b>{ }</b> are located.

Furthermore, you can also format how you want the numbers to appear, as shown below:

In [47]:
print('Area of my rectangle is {:.5f} cm^2 and the area of my circle is {:.2f} cm^2'. \
      format(calculate_rectangle_area(4,6), calculate_circle_area(4)))

Area of my rectangle is 24.00000 cm^2 and the area of my circle is 50.27 cm^2


Where the <b><code>:.5f</code></b> states that you want to print that variable with $5$ decimal numbers. Similarly, <b><code>:.2f</code></b> rounds the number to $2$ decimal numbers. More information on this in Section 3.1, in Notebook 3.

### Documenting functions
We have now successfully created a function that computes the area of a circle and the area of a rectangle. You could send this code to fellow students, but maybe they wouldn't know how to use them. This is where a <b>docstring</b> comes in handy. This is a string specified in the beginning of the function body which states information about the function. A lot of built-in Python functions also have docstrings, which is really useful when you're trying to understand how to use a function.

Add a description to the <code>calculate_circle_area(r)</code> function below, as a docstring.<br> 

In [48]:
def calculate_circle_area(r):
    '''This function calculate the area of a circle with radius r '''
    pi_circle = 3.141592653589793
    area = pi_circle*(r**2)
    return area

As you can see, nothing happened. But, if we now call the function like this:
```
calculate_circle_area?
```
or:
```
help(calculate_circle_area)
```
we should see the description (docstring) of the function. Try yourself below:

In [49]:
help(calculate_circle_area)


Help on function calculate_circle_area in module __main__:

calculate_circle_area(r)
    This function calculate the area of a circle with radius r



Now this isn't of much use when you work on your own code, unless you are very forgetful or have to write large programs.
But if you work using other people's code, this is really useful, as it helps you figure out how to use each function.

#### When to write or use a function?

You can use functions in your code when you have a specific piece of code that you need to use multiple times (e.g.: plotting something).<br><br>Often you will find you want to use an output from a function later on. To do this, you can assign a function to a variable. Let's say I want to use the area of a circle in a later calculation. Then you can store it in a variable like this:

In [50]:
Circle_Area = calculate_circle_area(4)
# here, we stored the area of a circle that has a radius 4 
# in the variable 'Circle_Area'

Nothing happens, but the value of <b><code>calculate_circle_area(4)</code></b> is now stored in the variable <b><code>Circle_Area</code></b>. See below:

In [51]:
print(Circle_Area)

50.26548245743669


We can see that the value was indeed stored.

::: {warning}
Variables that are defined inside a function body can <b>NOT</b> be called from outside of the function. These variables are called <b>local variables</b>.
:::

Take the variable **`pi_circle`** that we defined in the function **`calculate_circle_area()`**. If we try to print it:

In [52]:
print(pi_circle)

NameError: name 'pi_circle' is not defined

See, it doesn't work!
The error you get: <i>name 'pi_circle' is not defined</i>, means that you tried to call a variable that does not exist.

## Additional study material:

* Official Python Documentation - https://docs.python.org/3/tutorial/introduction.html
* https://realpython.com/python-variables/
* Think Python (2nd ed.) - Section 2

* Official Python Documentation - https://docs.python.org/3/library/operator.html
* https://realpython.com/python-operators-expressions/
* Think Python (2nd ed.) - Sections 2 and 5

* Official Python Documentation - https://docs.python.org/3/tutorial/controlflow.html#defining-functions
* https://realpython.com/defining-your-own-python-function/
* Think Python (2nd ed.) - Sections 3, 6 and 16

### After this Notebook you should be able to:

- understand why you need to learn Python
- create and re-assign new variables
- determine the type of a variable using `type()`
- slice strings
- perform simple math using arithmetic operators
- compare two or more variables
- check if a value, or element, exists in an object
- define a function, including its docstring